prevector: fix 2 bugs in currently unreached code paths

kazcw opened the pull request:
It looks like maybe this swap method is from an earlier iteration of prevector that used the LSB of size as a "direct/indirect" tag. The bad path isn't ever hit because in all current instances of swapping two prevectors, one is newly value-initialized and thus has an even size (0).

kazcw committed:
prevector: destroy elements only via erase()

Fixes a bug in which pop_back did not call the deleted item's destructor.

Using the most general erase() implementation to implement all the others
prevents similar bugs because the coupling between deallocation and destructor
invocation only needs to be maintained in one place.
Also reduces duplication of complex memmove logic.

kazcw committed:
prevector::swap: fix (unreached) data corruption

swap was using an incorrect condition to determine when to apply an optimization
(not swapping the full direct[] when swapping two indirect prevectors).

Rather than correct the optimization I'm removing it for simplicity. Removing
this optimization minutely improves performance in the typical (currently only)
usage of member swap(), which is swapping with a freshly value-initialized
object.

laanwj commented:
Concept ACK.
I think we need a unit test that fail before this, and succeed after this.

sipa commented:
Nice catch, thanks.

Code review ACK, but agree that a test for this behaviour would be welcome.

kazcw committed:
test prevector::swap

- add a swap operation to prevector tests (fails due to broken prevector::swap)
- fix 2 prevector test operation conditions that were impossible

laanwj commented:
Thanks for adding the test. I verified that it passes with, and fails without a7af72a.
tACK a7af72a
